File: frontend-api/src/main/java/no/nav/melosys/tjenester/gui/dto/trygdeavtale/TrygdeavtaleInfoDto.java
package no.nav.melosys.tjenester.gui.dto.trygdeavtale;

import java.time.LocalDate;
import java.util.List;
import java.util.Map;
import java.util.function.Predicate;

import com.fasterxml.jackson.annotation.JsonInclude;
import no.nav.melosys.domain.behandlingsgrunnlag.data.MedfolgendeFamilie;
import no.nav.melosys.domain.behandlingsgrunnlag.data.Periode;
import no.nav.melosys.domain.kodeverk.Trygdeavtale_myndighetsland;

public record TrygdeavtaleInfoDto(
    String aktoerId,
    String behandlingstema,
    String behandlingstype,
    boolean redigerbart,
    LocalDate periodeFom,
    LocalDate periodeTom,
    @JsonInclude(JsonInclude.Include.NON_NULL) Trygdeavtale_myndighetsland soeknadsland,
    List<VirksomhetDto> virksomheter, List<FamilieDto> barn,
    FamilieDto ektefelleSamboer,
    String innledningFritekst,
    String begrunnelseFritekst,
    String nyVurderingBakgrunn) {

    public TrygdeavtaleInfoDto(String aktoerId, String behandlingstema, String behandlingstype, boolean redigerbart, Periode periode, Trygdeavtale_myndighetsland soeknadsland, Map<String, String> virksomheter, List<MedfolgendeFamilie> familie, String innledingFritekst, String begrunnelseFritekst, String nyVurderingBakgrunn) {
        this(
            aktoerId,
            behandlingstema,
            behandlingstype,
            redigerbart,
            periode.getFom(),
            periode.getTom(),
            soeknadsland,
            mapVirksomheter(virksomheter),
            filtrerOgMapFamilie(familie, MedfolgendeFamilie::erBarn),
            filtrerOgMapFamilie(familie, MedfolgendeFamilie::erEktefelleSamboer)
                .stream().findFirst().orElse(null),
            innledingFritekst,
            begrunnelseFritekst,
            nyVurderingBakgrunn
        );
    }

    public static List<VirksomhetDto> mapVirksomheter(Map<String, String> virksomheter) {
        return virksomheter.entrySet().stream()
            .map(virksomhet -> new VirksomhetDto(virksomhet.getKey(), virksomhet.getValue()))
            .toList();
    }

    public static List<FamilieDto> filtrerOgMapFamilie(List<MedfolgendeFamilie> familie, Predicate<MedfolgendeFamilie> filterfunksjon) {
        return familie.stream()
            .filter(filterfunksjon)
            .map(familiemedlem -> new FamilieDto(familiemedlem.getUuid(), familiemedlem.getFnr(), familiemedlem.getNavn()))
            .toList();
    }
}

File: frontend-api/src/main/kotlin/no/nav/melosys/tjenester/gui/dto/trygdeavtale/TrygdeavtaleInfoDto.kt
package no.nav.melosys.tjenester.gui.dto.trygdeavtale

import com.fasterxml.jackson.annotation.JsonInclude
import no.nav.melosys.domain.kodeverk.Trygdeavtale_myndighetsland
import no.nav.melosys.domain.mottatteopplysninger.data.MedfolgendeFamilie
import no.nav.melosys.domain.mottatteopplysninger.data.Periode
import java.time.LocalDate
import java.util.function.Function
import java.util.function.Predicate

@JvmRecord
data class TrygdeavtaleInfoDto(
    @JvmField val aktoerId: String?,
    @JvmField val behandlingstema: String?,
    @JvmField val behandlingstype: String?,
    @JvmField val redigerbart: Boolean,
    @JvmField val periodeFom: LocalDate,
    @JvmField val periodeTom: LocalDate,
    @JvmField @field:JsonInclude(JsonInclude.Include.NON_NULL) @param:JsonInclude(JsonInclude.Include.NON_NULL) val soeknadsland: Trygdeavtale_myndighetsland?,
    val virksomheter: List<VirksomhetDto?>, val barn: List<FamilieDto?>,
    val ektefelleSamboer: FamilieDto?,
    @JvmField val innledningFritekst: String?,
    @JvmField val begrunnelseFritekst: String?,
    @JvmField val nyVurderingBakgrunn: String?
) {
    constructor(
        aktoerId: String?,
        behandlingstema: String?,
        behandlingstype: String?,
        redigerbart: Boolean,
        periode: Periode,
        soeknadsland: Trygdeavtale_myndighetsland?,
        virksomheter: Map<String?, String?>,
        familie: List<MedfolgendeFamilie>,
        innledingFritekst: String?,
        begrunnelseFritekst: String?,
        nyVurderingBakgrunn: String?
    ) : this(
        aktoerId,
        behandlingstema,
        behandlingstype,
        redigerbart,
        periode.fom,
        periode.tom,
        soeknadsland,
        mapVirksomheter(virksomheter),
        filtrerOgMapFamilie(familie) { obj: MedfolgendeFamilie -> obj.erBarn() },
        filtrerOgMapFamilie(familie) { obj: MedfolgendeFamilie -> obj.erEktefelleSamboer() }
            .stream().findFirst().orElse(null),
        innledingFritekst,
        begrunnelseFritekst,
        nyVurderingBakgrunn
    )

    companion object {
        fun mapVirksomheter(virksomheter: Map<String?, String?>): List<VirksomhetDto?> {
            return virksomheter.entries.stream()
                .map { a -> VirksomhetDto(a.key, a.value) }
                .toList()
        }

        fun filtrerOgMapFamilie(
            familie: List<MedfolgendeFamilie>,
            filterfunksjon: Predicate<MedfolgendeFamilie>?
        ): List<FamilieDto?> {
            return familie.stream()
                .filter(filterfunksjon)
                .map { familiemedlem: MedfolgendeFamilie ->
                    FamilieDto(
                        familiemedlem.uuid,
                        familiemedlem.fnr,
                        familiemedlem.navn
                    )
                }
                .toList()
        }
    }
}
